### PR TITLE
i2264: write out barman status periodically

### PR DESCRIPTION
**Pull request by richfitz**
Also includes VIMC-2116 (#101)

Sorry for the size of this:

* Remove the flask app from the container
* replaced cron with `yacron` - this has the advantage of easy logging and easier configuration. This change will make adding a periodic base backup much easier in future ([VIMC-2206](https://vimc.myjetbrains.com/youtrack/issue/VIMC-2209))
* A new volume `barman_metrics` where metrics will be written to periodically. The application that @littledice is working on can mount that `:ro` to pick up metrics
* Simplify the initial backup logic - we now do an initial backup if one does not exist (i.e., on the first deploy of barman)
* Start all barman services as the container comes up - this and the initial backup changes above make restarting trivial

Metrics turn up as `metrics.json` in the metrics volume. An example output looks like:

```json
{
  "data": {
    "barman_time_since_last_backup_days": 0.0028623128356481484,
    "barman_check_replication_slot": 1,
    "barman_available_backups": 1,
    "barman_time_of_last_backup": 1539158454,
    "barman_check_archiver_errors": 1,
    "barman_check_failed_backups": 1,
    "barman_check_pg_basebackup_compatible": 1,
    "barman_check_receive_wal_running": 1,
    "barman_pg_version": "10.3",
    "barman_time_since_last_backup_minutes": 4.121730483333334,
    "barman_check_wal_level": 1,
    "barman_check_backup_maximum_age": 1,
    "barman_check_pg_basebackup": 1,
    "barman_check_directories": 1,
    "barman_time_since_last_backup_hours": 0.06869550805555556,
    "barman_check_PostgreSQL_streaming": 1,
    "barman_check_pg_receivexlog": 1,
    "barman_data_size": 35651584,
    "barman_check_pg_receivexlog_compatible": 1,
    "barman_time_since_last_backup_seconds": 247.303829,
    "barman_check_retention_policy_settings": 1,
    "barman_check_is_superuser": 1,
    "barman_check_compression_settings": 1,
    "barman_check_pg_basebackup_supports_tablespaces_mapping": 1,
    "barman_ok": 1,
    "barman_check_minimum_redundancy_requirements": 1,
    "barman_check_PostgreSQL": 1,
    "barman_running": true
  },
  "utc_seconds": 1539158701.303853,
  "utc_string": "2018-10-10T08:05:01.303853+00:00"
}
```

### CODE REVIEW

File: backup/bin/barman-parse-metrics [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+import json
+import os
+import re
+import tempfile
+
+from datetime import datetime, timezone
+from dateutil import parser
+
+
+# These should come from Alex's module code
```
Comment (MartinEden) [inline backup/bin/barman-parse-metrics]: Outstanding TODO

File: backup/bin/barman-parse-metrics [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+def seconds_elapsed_since(timestamp):
+    if timestamp:
+        now = datetime.now()
+        return (now - timestamp).total_seconds()
+    else:
+        return None
+
+
+def clean_key(k, prefix):
+    return prefix + re.sub("[^A-Za-z0-9]", "_", k)
```
Comment (MartinEden) [inline backup/bin/barman-parse-metrics]: Maybe change this to `[^A-Za-z0-9]+`, so that multiple illegal characters collapse to 1 underscore?

File: backup/bin/barman-parse-metrics [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        return {}
+
+
+def parse_status(path):
+    check = barman_output_as_dict(read_lines(path + '/check.txt'))
+    status = barman_output_as_dict(read_lines(path + '/status.txt'))
+    ret = {clean_key(k, "barman_check_"): int(v.startswith("OK"))
+           for (k, v) in check.items()}
+    ret["barman_ok"] = all(x == 1 for x in ret.values())
+    ret["barman_running"] = 1
```
[the remark this answers is not in the document]
Comment (MartinEden) [reply backup/bin/barman-parse-metrics]: We were trying to distinguish between three states, but this may no longer be relevant:

* Barman is not running
* Barman is running, but an error occurred getting status from it
* Barman is running and we were able to get the status

File: backup/bin/barman-parse-metrics [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    ret["barman_data_size"] = parse_size(status["Current data size"])
+    ret["barman_pg_version"] = status["PostgreSQL version"]
```
[the remark this answers is not in the document]
Comment (MartinEden) [reply backup/bin/barman-parse-metrics]: I was going to leave it until/unless the version string was ever anything non-numeric

File: backup/bin/barman-parse-metrics [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        return f.read()
+
+
+if __name__ == "__main__":
+    args = docopt.docopt(__doc__)
+    path = args['<path>']
+    try:
+        dat = parse_status(path)
+    except:
+        dat = {"barman_running": 0}
```
Comment (MartinEden) [inline backup/bin/barman-parse-metrics]: Should we log the exception? At least to stdout?

File: backup/bin/barman-save-metrics [hunk 6]
```diff
@@ -0,0 +1,23 @@
+#!/usr/bin/env bash
+
+# NOTE: not run with '-e' because we need to continue even if the
+# check/status commands fail.
```
Comment (MartinEden) [inline backup/bin/barman-save-metrics]: Why not -e except for those two lines?

File: backup/bin/setup-barman [hunk 7]
```diff
@@ -109,21 +106,15 @@ def initial_backup(conf):
...
-    subprocess.run(["service", "cron", "start"], check=True)
+        base_dir = '/var/lib/barman/{}/base'.format(nm)
+        if os.path.exists(base_dir) and os.listdir(base_dir):
+            print("Base backup already exists for {}".format(nm))
+        else:
+            force_archive(nm)
+            print("Running initial backup of {}".format(nm), flush=True)
+            subprocess.run(["barman", "backup", nm], check=True)
+            print("Checking barman status of {}".format(nm), flush=True)
+            subprocess.run(["barman", "check", nm], check=True)
```
Comment (MartinEden) [inline backup/bin/setup-barman]: I see, so we only do a base backup if we haven't already done one. And this means we don't need the `no-initial-backup` flag when we call this from the docker entrypoint.

File: backup/bin/setup-barman [hunk 8]
```diff
@@ -109,21 +106,15 @@ def initial_backup(conf):
...
     for x in conf:
         nm = x['configuration_name']
-        force_archive(nm)
-        print("Running initial backup of {}".format(nm), flush=True)
-        subprocess.run(["barman", "backup", nm], check=True)
-        print("Checking barman status of {}".format(nm), flush=True)
-        subprocess.run(["barman", "check", nm], check=True)
-
-
-def setup_cron(conf):
```
Comment (MartinEden) [inline backup/bin/setup-barman]: So we no longer need this because the schedule.yaml file for yacron handles this now.

File: backup/bin/barman-parse-metrics (hunk 4 above)
[the remark this answers is not in the document]
Comment (richfitz) [reply backup/bin/barman-parse-metrics]: It's not a problem until after version 10.9 as it'll be monotonic at least, but at 10.10 we will see a "drop"

File: backup/bin/barman-save-metrics (hunk 6 above)
Comment (richfitz) [reply backup/bin/barman-save-metrics]: Doing that makes the script look a bit weird because we have set -e changing 3 times (top, above the `barman check` line, below the `barman status` line) but the lines that can cause set -e to throw are only `mkdir -p` (can't really fail), `rm -f` (can't fail), a test that's designed to exit and the final metrics parsing

File: backup/bin/setup-barman (hunk 7 above)
Comment (richfitz) [reply backup/bin/setup-barman]: Yeah, this way felt simplest to me. It doesn't allow for deploying a barman instance that just streams with no base backup but I don't think that's a useful thing to have really

File: backup/bin/setup-barman (hunk 8 above)
Comment (richfitz) [reply backup/bin/setup-barman]: yes, that's right

File: backup/bin/barman-save-metrics (hunk 6 above)
Comment (richfitz) [reply backup/bin/barman-save-metrics]: Actually, the last line should always be run too

File: backup/bin/barman-parse-metrics (hunk 3 above)
[the remark this answers is not in the document]
Comment (richfitz) [reply backup/bin/barman-parse-metrics]: ok, with the changes now we'll have:

`barman_ok` - top level warning signal - everything is ok
`barman_metrics_ok` - there was any error running the metrics command
`barman_running` - barman responded as expected

We can refine later

File: backup/Dockerfile [hunk 9]
```diff
@@ -21,6 +18,16 @@ RUN wget --quiet -O - https://www.postgresql.org/media/keys/ACCC4CF8.asc | \
                 barman \
                 postgresql-client-10
 
+RUN pip3 install \
+        docopt \
+        yacron
+
+RUN git clone https://github.com/vimc/montagu-metrics-py /tmp/metrics_utils && \
+        git -C /tmp/metrics_utils reset --hard 4b2ef9b && \
```
Comment (MartinEden) [inline backup/Dockerfile]: Should pull out this version to an env variable or similar?

File: backup/Dockerfile [hunk 10]
```diff
@@ -21,23 +18,24 @@ RUN wget --quiet -O - https://www.postgresql.org/media/keys/ACCC4CF8.asc | \
                 barman \
                 postgresql-client-10
 
+RUN pip3 install \
+        docopt \
+        yacron
+
+RUN git clone https://github.com/vimc/montagu-metrics-py /tmp/metrics_utils && \
```
Comment (richfitz) [inline backup/Dockerfile]: We're having enough fun with submodules that I though I'd try this way.

File: backup/Dockerfile [hunk 11]
```diff
@@ -21,23 +18,24 @@ RUN wget --quiet -O - https://www.postgresql.org/media/keys/ACCC4CF8.asc | \
...
+        git -C /tmp/metrics_utils reset --hard 4b2ef9b && \
```
Comment (littledice) [inline backup/Dockerfile]: Why is this here?

Comment (richfitz) [reply backup/Dockerfile]: To get the metrics code into the docker image. The other way is to add a submodule, do a `COPY` in the dockerfile and manage versions that way. Given submodules are a source of grief and we never need that code in the non-docker part of this, I thought we could try this approach

**Comment by richfitz** [comment]
> It looks like you want the metrics library to be published on PyPi. Is this something to add to the backlog?

Not incredibly - it might be worthwhile but it would (from the look of it) require even further refactoring of the metrics code

**Comment by MartinEden** [comment]
Waiting to merge https://github.com/vimc/cached-metrics/pull/1 at the same time.
